include/cuda/memory_resource

```diff
@@ -0,0 +1,318 @@
+   * resource can be deallocated from the other and vice versa.
+   *
+   * @param __other The other resource to compare against
+   */
+  bool is_equal(memory_resource const& __other) const noexcept {
+    return do_is_equal(__other);
+  }
+
+private:
+  virtual void *do_allocate(std::size_t __bytes, std::size_t __alignment) = 0;
+
+  virtual void do_deallocate(void *__p, std::size_t __bytes,
+                             std::size_t __alignment) = 0;
+
+  // Default to identity comparison
+  virtual bool do_is_equal(memory_resource const &__other) const noexcept{
+      return this == &__other;
+  }
+
+  // Default to return default constructed Context object
+  virtual context do_get_context() const noexcept { return _Context{}; }
+};
+
+/**
+ * @brief Abstract interface for CUDA stream-ordered memory allocation.
+ *
+ * "Stream-ordered memory allocation" extends the CUDA programming model to
+ * include memory allocation as stream-ordered operations.
+ *
+ * Allocating on stream `s0` returns memory that is valid to access immediately
+ * only on `s0`. Accessing it on any other stream (or the host) first requires
+ * synchronization with `s0`, otherwise behavior is undefined.
+ *
+ * Deallocating memory on stream `s1` indicates that it is valid to reuse the
+ * deallocated memory immediately for another allocation on `s1`.
+ *
+ * Memory may be allocated and deallocated on different streams, `s0` and `s1`
+ * respectively, but requires synchronization between `s0` and `s1` before the
+ * deallocation occurs.
+ *
+ * @tparam _Kind The `memory_kind` of the allocated memory.
+ */
+template <memory_kind _Kind>
+class stream_ordered_memory_resource : public memory_resource<_Kind /* default context */> {
+public:
+  using memory_resource<_Kind>::kind;
+  using memory_resource<_Kind>::default_alignment;
+
+  /**
+   * @brief Allocates storage of size at least `__bytes` bytes in stream order
+   * on `__stream`.
+   *
+   * The returned storage is aligned to `default_alignment`.
+   *
+   * The returned storage may be used immediately only on `__stream`. Accessing
+   * it on any other stream (or the host) requires first synchronizing with
+   * `__stream`.
+   *
+   * @throws If the storage of the requested size cannot be obtained.
+   *
+   * @param __bytes The size in bytes of the allocation.
+   * @param __stream The stream on which to perform the allocation.
+   * @return Pointer to the requested storage.
+   */
+  void *allocate_async(std::size_t __bytes, stream_view __stream) {
+    return do_allocate_async(__bytes, default_alignment, __stream);
+  }
+
+  /**
+   * @brief Allocates storage of size at least `__bytes` bytes in stream order
+   * on `__stream`.
+   *
+   * The returned storage is aligned to the specified `__alignment` if such
+   * alignment is supported.
+   *
+   * The returned storage may be used immediately only on `__stream`. Using it
+   * on any other stream (or the host) requires first synchronizing with
+   * `__stream`.
```

Comment:
Clarify that synchronization needs to happen before the access, not before the allocation.

Reply:
Use "happens-before/happens-after" if possible.